docker/scripts/p4runtime_switch.py

```diff
@@ -0,0 +1,321 @@
+        if self.enable_debugger:
+            args.append("--debugger")
+        if self.log_console:
+            args.append("--log-console")
+        if self.thrift_port:
+            args.append('--thrift-port ' + str(self.thrift_port))
+        if self.grpc_port:
+            args.append("-- --grpc-server-addr 0.0.0.0:" + str(self.grpc_port))
+        cmd = ' '.join(args)
+        info(cmd + "\n")
+
+
+        pid = None
+        with tempfile.NamedTemporaryFile() as f:
+            self.cmd(cmd + ' >' + self.log_file + ' 2>&1 & echo $! >> ' + f.name)
+            pid = int(f.read())
+        debug("P4 switch {} PID is {}.\n".format(self.name, pid))
+        if not self.check_switch_started(pid):
+            error("P4 switch {} did not start correctly.\n".format(self.name))
+            exit(1)
+        info("P4 switch {} has been started.\n".format(self.name))
+
+        if self.start_controller:
+            self.sw_conn = p4runtime_lib.bmv2.Bmv2SwitchConnection(
+                    name=self.name,
+                    address='127.0.0.1:' + str(self.grpc_port),
+                    device_id=self.device_id,
+                    proto_dump_file='/tmp/p4app-logs/' + self.name + '-p4runtime-requests.txt')
+
+            try:
+                self.sw_conn.MasterArbitrationUpdate()
+            except grpc.RpcError as e:
+                printGrpcError(e)
+
+            if self.p4info_path:
+                self.loadP4Info()
+            self.loadJSON()
+
+    def loadP4Info(self):
+        self.p4info_helper = p4runtime_lib.helper.P4InfoHelper(self.p4info_path)
+
+    def loadJSON(self):
+        try:
+            self.sw_conn.SetForwardingPipelineConfig(p4info=self.p4info_helper.p4info,
+                    bmv2_json_file_path=self.json_path)
+        except grpc.RpcError as e:
+            printGrpcError(e)
+
+    def loadConf(self, sw_conf_or_filename):
+        if isinstance(sw_conf_or_filename, dict):
+            sw_conf = sw_conf_or_filename
+        else:
+            conf_path = os.path.join('/p4app', sw_conf_or_filename)
+            with open(conf_path, 'r') as f:
+                sw_conf = json_load_byteified(f)
+
+        if 'p4info' in sw_conf:
+            info('Using P4Info file %s...' % sw_conf['p4info'])
+            self.p4info_path = os.path.join('/tmp/p4app-logs/', sw_conf['p4info'])
+            self.loadP4Info()
+
+        assert sw_conf['target'] == 'bmv2'
+
+        if 'bmv2_json' in sw_conf:
+            info("Setting pipeline config (%s)..." % sw_conf['bmv2_json'])
+            self.json_path = os.path.join('/tmp/p4app-logs/', sw_conf['bmv2_json'])
+            self.loadJSON()
+
+        if 'table_entries' in sw_conf:
+            info("Inserting %d table entries..." % len(sw_conf['table_entries']))
+            for entry in sw_conf['table_entries']:
+                info(tableEntryToString(entry))
+                self.insertTableEntry(entry)
+
+    def insertTableEntry(self, entry=None,
```

Comment:
obviously a lot of P4Runtime functionalities are missing from this class. I imagine support can be added as time goes by, but I believe that the app should also have access to the underlying P4Runtime connection.